File: DFS/Basic/200. Number of Islands/200. Number of Islands.cpp
class Solution {
    int m, n;
    int res = 0;
    vector<vector<int>> visited;
    vector<pair<int,int>> directions{{0,1},{0,-1},{1,0},{-1,0}};
public:
    int numIslands(vector<vector<char>>& grid) {
        
        m = grid.size();
        n = grid[0].size();
        visited.resize(m, vector<int>(n, 0));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (visited[i][j] || grid[i][j] == '0') continue;
                dfs(grid,i,j);
                res++;
            }
        }
        return res;
    }
    
    void dfs(vector<vector<char>>& grid, int i, int j) {
        
        for (int k = 0; k < 4; k++) {
            int x = i + directions[k].first;
            int y = j + directions[k].second;
            if (x < 0 || x >= m || y < 0 || y >= n) continue;
            if (!visited[x][y] && grid[i][j] == '1') {
                visited[x][y] = 1;
                dfs(grid, x, y);
            }
        }
        return;
    }
};

File: DFS/Basics/200. Number of Islands/200. Number of Islands.java
class Solution {
    char graph[][];
    int dirs[][] = {{0,1},{0,-1},{1,0},{-1,0}};
    public int numIslands(char[][] grid) {
        
        graph = grid;
        int count = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] != '1') continue;
                dfs(i, j);
                count++;
            }
        }
        return count;
    }
    
    public void dfs(int r, int c) {
        
        for (int k = 0; k < 4; k++) {
            int x = r + dirs[k][0];
            int y = c + dirs[k][1];
            if (x < 0 || x >= graph.length || y < 0 || y >= graph[0].length) continue;
            if (graph[x][y] != '1') continue;
            graph[x][y] = '2';
            dfs(x,y);
        }
    }
}